README.md
### Overview

The R package **tagr** implements a non-parametric bootstrap to estimate confidence intervals for Chapman/Petersen estimates of abundance


### Examples

To learn more about tagr, start with the vignettes:

```S
browseVignettes(package = "tagr")
```




NEWS
CHANGES in tagr dev VERSION 0.1.8

o adding tags available for each bootstrap replicate to the output of 
bootstrap.mrelease (request of Lucy Robinson)

CHANGES in tagr VERSION 0.1.7

o multi_release bootstrap method (bootstrap.mrelease) now has option to choose
whether population size will be estimated for bootstrap replicates with zero
recaptures (request of Keith Reid).

CHANGES in tagr VERSION 0.1.5

o bug fix, weight based Chapman bootstrap using multi_release function was instead
providing the overall estimate

CHANGES in tagr VERSION 0.1.4

o bug fix, weight based Chapman estimate using multi_release function was instead
providing the overall estimate

CHANGES in tagr VERSION 0.1.3

o bug fix, weight based Chapman estimate using multi_release function was not
provided with the catch weight by haul. The first column of haul data must now
be the total catch weight (tagged+untagged). The subsequent columns still 
represent the numbers of recaptured fish by release cohort

CHANGES in tagr VERSION 0.1.2

o added TagsAvailable to obj list in single release

o bug fix in check_type

CHANGES in tagr VERSION 0.1.1

o new bootstrap method for 'mrelease' objects type 1 fisheries

CHANGES in tagr VERSION 0.1.0

o new multi_release function

o upgrade to roxygen version 6.0

o option to calculate population size with zero recaptures  

CHANGES in tagr VERSION 0.0.15

o update of documentation

CHANGES in tagr VERSION 0.0.14

o ilogit transformation functions moved to 'tagsurviv' package

o upgrade to roxygen2 version 5.0

CHANGES in tagr VERSION 0.0.13

o correction to chapman_n, was adding 1 rather than subtracting 1 
(thanks to Kenji Taki)

o inst/doc folder with vignette added to github

CHANGES in tagr VERSION 0.0.11

o adding the option of including the mean weight of a single fish to
chapman_wt and associated functions

o correcting a minor error in chapman_wt variance

o added option to specify the quantiles probabilities for summary.bsamples
method

o added bootstrapped standard error and coefficient of variation to 
summary.bsamples method

o added coefficient of variation to summary.srelease

o additional tests for single_release and associated functions

CHANGES in tagr VERSION 0.0.9

o uploaded to Github https://github.com/paulburch/tagr

CHANGES in tagr VERSION 0.0.7

o variance now calculated for "Chapman" method of single_release when
catch is in weight

o improvments to documentation for single_release bootstrap method and 
the package

tests/testthat/test-srelease.r
## tests for single release tag recapture models

# load packages for writing new tests
# library(tagr)
# library(testthat)

## test check_srelease_inputs
test_that("test srelease inputs", {
   expect_warning(check_srelease_inputs(-1, 10, 1), "releases, catch and recaptures must not be negative")
})


## test estimates of population size
test_that("test srelease N_hat", {
  expect_that(chapman_n(tags=100, catch=100, recaps=10)[["N_hat"]], equals(928.3636, tolerance=1e-4)) 
  expect_that(chapman_wt(tags=100, catch=100, recaps=10, mean_wt=5)[["N_hat"]], equals(959.0909, tolerance=1e-4))
  expect_that(petersen(tags=100, catch=100, recaps=10)[["N_hat"]], equals(1000, tolerance=1e-4)) 
  })

## test variances of population size
test_that("test srelease var_N", {
  expect_that(chapman_n(tags=100, catch=100, recaps=10)[["var_N"]], equals(56906.4, tolerance=1e-4)) 
  expect_that(chapman_wt(tags=100, catch=100, recaps=10, mean_wt=5)[["var_N"]], equals(32866.74, tolerance=1e-4))
  expect_that(petersen(tags=100, catch=100, recaps=10)[["var_N"]], equals(90000, tolerance=1e-4)) 
})


R/tagr.R
#' tagr: A package for tagging models to estimate abundance
#'
#' The tagr package provides methods for tag-based estimates of 
#' abundance 
#' 
#' @section Single release tag recapture models:
#' Implementation of Petersen and Chapman estimators (Seber 1982)
#' along with bootstrapped confidence intervals
#'
#' @section Vignettes:
#' To learn more about tagr, start with the vignettes:
#' \code{browseVignettes(package = "tagr")}
#'
#' @docType package
#' @importFrom stats quantile rbinom sd
#' @name tagr
NULL
#> NULL


#' Numbers of carp and those with tags captured in 2014/15
#'
#' A dataset containing fishing events for carp in Lake 
#' Sorell. The variables are as follows:
#'
#' @format A data frame with 195 rows and 3 variables:
#' \itemize{
#'   \item day: day of fishing event starting 1st July 2014
#'   \item catch: total carp caught
#'   \item tagged: number of tagged carp caught
#' }
"carphauls"

#' Tag release and recapture data from Lake Sorell
#'
#' A dataset containing tag release and recapture information for carp 
#' released in January 2012. All fish were double tagged with T-bar tags. 
#' The variables are as follows:
#'
#' @format A data frame with 803 rows and 9 variables:
#' \itemize{
#'   \item release: either first (1) or second (2)
#'   \item tags: number of tags on recaptured fish (either 1, 2), NA denotes fish still at liberty
#'   \item date: date of recapture, NA denotes fish still at liberty
#'   \item season: season (July to June) of recapture, NA denotes fish still at liberty
#' }
"carptags"


R/checks.R
## check function inputs
## investigate further 
## http://adv-r.had.co.nz/Exceptions-Debugging.html#condition-handling

check_srelease_inputs <- function(tags, catch, recaps, zeroes=FALSE){
  ## define the check variable
  check <- TRUE
  ## check if any inputs are < 0
  if(any(tags<=0, catch<=0, recaps<0)){
    check <- FALSE
    warning("releases, catch and recaptures must not be negative")
  }
  ## check if recaptures are greater than releases
  if(tags < recaps){ 
    check <- FALSE
    warning("more tags have been recaptured than were released")
  }
  ## check is recaptures are greater than catch
  if(catch < recaps){
    check <- FALSE
    warning("more tags have been recaptured than were released")
  }
  ## if there are no recaptures we don't estimate population size
  if(recaps==0){
    if(!zeroes){
      warning("there are zero recapures population size not estimated")
      check <- FALSE
    }else{
      warning("there are zero recapures and population size has been estimated")
    }
    ## now check that the adjusted releases are > 0
  }
  if(tags <= 0){
    ## alternately N_hat could be zero
    check <- FALSE
    warning("there are zero releases population size not estimated")
  }
  ## return check
  check
}

check_mrelease_inputs <- function(tags, hauls, pars, zeroes=FALSE){
  ## define the check variable
  check <- TRUE
  ## add checks
  ## check the tag data first
  ## check each row of releases is < total recaptures
  # for(i in 1:nrow(tags)){
  #   if(tags[i, 1] < sum(tags[i, 2:ncol(tags)]))
  #     warning(paste0("More tags have been recaptured than were released in year", i))
  #   check <- FALSE
  # }
  ## check recaptures in the haul data matches the tag data
  
  
  ## return check
  check
}


R/generics.R
## ************************************************ ##
## New S3 Methods

#' Bootstrap method for single release tag-return studies
#'
#' Bootstrap method for single release tag-return studies
#' 
#' Estimate confidence intervals using a non-parametric bootstrap. This method
#' incoporates uncertainty in tag-induced mortality, natural mortality and tag 
#' shedding as a series of Bernoulli trials. @seealso \code{\link{single_release}}
#' for more details.
#' @param x an object of class srelease
#' @param nboot number of bootstrap samples
#' @param boot_zeroes should bootstrap replicates with zero tag-recaptures be
#' used to estimate population size (default=TRUE)
#' @param ... additional parameters
#' @aliases bootstrap.srelease
#' @export
bootstrap <- function(x, nboot, boot_zeroes=TRUE, ...)
  UseMethod("bootstrap")


R/estimators.R
## tag-based estimators of population size 

#' Petersen estimators
#' 
#' Single tag recapture estimates Petersen, Lincoln Petersen estimators
#' @param tags number of marked animals released (note these must be whole 
#' numbers for the bootstrap to work)
#' @param catch vector of the number or weight of animals captured and checked 
#' for tags by haul on the second survey
#' @param recaps vector of number of marked animals recaptured by haul
#' @param mean_wt mean weight of a single fish (optional argument for Chapman
#'  weight method)
#' @param check type of input checks "srelease" (default= single release checks)
#' other options are "mrelease" (multiple release checks) and "zero_recaps"
#' (include 
#' @name petersen
NULL
#> NULL


#' @export
#' @rdname petersen
chapman_n <- function(tags, catch, recaps, check_type="srelease"){
  ## check function inputs
  switch(check_type,
         srelease = {check <- check_srelease_inputs(tags, catch, recaps)},
         mrelease = {check <- check_mrelease_inputs(tags, catch, recaps)})
  ## if check passes calculate population size
  if(check){
    ## calculate the estimate
    N_hat <- (((tags + 1)*(catch + 1))/(recaps + 1)) - 1 
    ## calculate the variance
    var_N <- ((tags + 1)*(catch + 1)*(tags - recaps)*(catch - recaps))/
      (((recaps + 1)^2)*(recaps + 2))
    obj <- c(N_hat, var_N)
  }else{
    obj <- c(NA, NA)
  }
  names(obj) <- c("N_hat", "var_N")
  obj
}

#' @export
#' @rdname petersen
chapman_wt <- function(tags, catch, recaps, mean_wt=0, check="srelease"){
  ## check function inputs
  switch(check_type,
         srelease = {check <- check_srelease_inputs(tags, catch, recaps)},
         mrelease = {check <- check_mrelease_inputs(tags, catch, recaps)})
  ## if check passes calculate population size
  if(check){
    ## calculate the estimate
    N_hat <- (((tags + 1)*(catch + mean_wt))/(recaps + 1)) - mean_wt
    ## calculate the variance
    var_N <- ((tags + 1)*(catch + mean_wt)*(tags - recaps)*(catch - mean_wt*recaps))/
      (((recaps + 1)^2)*(recaps + 2))
    obj <- c(N_hat, var_N)
  }else{
    obj <- c(NA, NA)
  }
  names(obj) <- c("N_hat", "var_N")
  obj
}

#' @export
#' @rdname petersen
petersen <- function(tags, catch, recaps, check="srelease"){
  ## check function inputs
  switch(check_type,
         srelease = {check <- check_srelease_inputs(tags, catch, recaps)},
         mrelease = {check <- check_mrelease_inputs(tags, catch, recaps)})
  ## if check passes calculate population size
  if(check){
    ## calculate the estimate
    N_hat <- tags * catch / recaps 
    ## variance from Ricker 1975
    var_N <- (tags^2) * catch * (catch - recaps) / recaps^3 
    ## return the estimate and variance
    obj <- c(N_hat, var_N)
  }else{
    obj <- c(NA, NA)
  }
  names(obj) <- c("N_hat", "var_N")
  obj
}


vignettes/srelease.R
## ------------------------------------------------------------------------
library(tagr)
attach(carptags)
head(carptags)

## ------------------------------------------------------------------------
with(carptags, table(season, release))

## ------------------------------------------------------------------------
(rel1 <- nrow(carptags[carptags$release==1,]))
(rel2 <- nrow(carptags[carptags$release==2,]))


## ------------------------------------------------------------------------
(recaps1 <- with(carptags, table(season, release))[,1])
(recaps2 <- with(carptags, table(season, release))[,2])


## ------------------------------------------------------------------------
(rate1 <- (recaps1[1]/rel1))

## ------------------------------------------------------------------------
(rate2 <- (recaps2[1]/rel2))

## ------------------------------------------------------------------------
(rel1 - (rel1 * (recaps1[1] / (rel1 * (rate2))))) / (rel1+rel2)


## ------------------------------------------------------------------------
ls <- with(carptags, table(season, tags))
ls_prop <- ls[,1]/ls[,2]
cbind(ls, ls_prop)


## ------------------------------------------------------------------------
catches <- c(3181, 4377, 2417, 1253)


## ------------------------------------------------------------------------
tagged_catch <- with(carptags, (table(season)))


## ------------------------------------------------------------------------
prop <- tagged_catch/catches


## ------------------------------------------------------------------------
1 - (prop[3] / prop[2])


## ------------------------------------------------------------------------
1 - (prop[4] / prop[3])


## ------------------------------------------------------------------------
attach(carphauls)
head(carphauls)

## ------------------------------------------------------------------------
(prior_recaps <- recaps1[1:3] + recaps2[1:3])


## ------------------------------------------------------------------------
(recaps <- recaps1[4] + recaps2[4])
recaps

## ------------------------------------------------------------------------
sum(carphauls$tagged)

## ------------------------------------------------------------------------
set.seed(15)

## ------------------------------------------------------------------------
est_1 <- single_release(tags = (rel1 + rel2),
                        catch = carphauls$catch,
                        recaps = carphauls$tagged,
                        prior_recaps = prior_recaps,
                        method = "Chapman",
                        unit = "numbers",
                        type = 2,
                        tag_mort = 0.2,
                        nat_mort = 0.04,
                        chronic_mort = 0.17)

summary(est_1)


## ------------------------------------------------------------------------
ci_1 <- bootstrap(est_1, nboot=1000)
summary(ci_1)

## ------------------------------------------------------------------------
est_2 <- single_release(tags = (rel1 + rel2),
                        catch = carphauls$catch,
                        recaps = carphauls$tagged,
                        prior_recaps = prior_recaps,
                        method = "Chapman",
                        unit = "numbers",
                        type = 2,
                        tag_mort = 0.25,
                        nat_mort = 0.04,
                        chronic_mort = 0.17)

summary(est_2)


## ------------------------------------------------------------------------
ci_2 <- bootstrap(est_2, nboot=1000)
summary(ci_2)

## ------------------------------------------------------------------------
est_3 <- single_release(tags = (rel1 + rel2),
                        catch = carphauls$catch,
                        recaps = carphauls$tagged,
                        prior_recaps = prior_recaps,
                        method = "Chapman",
                        unit = "numbers",
                        type = 2,
                        tag_mort = 0.3,
                        nat_mort = 0.04,
                        chronic_mort = 0.17)

summary(est_3)


## ------------------------------------------------------------------------
ci_3 <- bootstrap(est_3, nboot=1000)
summary(ci_3)

## ------------------------------------------------------------------------
sessionInfo()



R/single_release.R
#' Single tag release estimate of population size
#' 
#' Estimate population size from single release tag data
#' 
#' For a Type 1 fishery the adjusted releases are calculate as follows
#' 
#' 1) Initial tag-induced mortality is applied
#' 
#' 2) Within season recaptures divided by the tag reporting rate are removed
#' 
#' 3) Natural mortality and chronic tag loss and tag-induced mortality is
#'  applied.
#' 
#' For a fishery that operates throughout the year (Ricker Type 2) fishing and 
#' natural mortality competed to deplete the tagged population. The adjusted 
#' releases are calculated as follows
#' 
#' 1) Initial tag-induced mortality is applied
#' 
#' 2) Half of the within season recaptures divided by the tag reporting rate are
#' removed 
#' 
#' 3) Half of the natural mortality and chronic tag loss and tag-induced 
#' mortality is applied
#' 
#' 4) The remaining within season recaptures divided by the tag reporting rate 
#' are removed 
#' 
#' 5) The remaining natural mortality and chronic tag loss and tag-induced 
#' mortality is applied.
#' @param tags number of marked animals released (note these must be whole 
#' numbers for the bootstrap to work)
#' @param catch vector of the number or weight of animals captured and checked 
#' for tags by haul on the second survey
#' @param recaps vector of number of marked animals recaptured by haul
#' @param mean_wt mean weight of a single fish (optional argument for Chapman
#'  weight method)
#' @param prior_recaps vector of subsequent recaptures (first year this 
#' represents within season recaptures)
#' @param method method
#' @param unit measurement unit (numbers, kg, tonnes)
#' @param type do we assume fishing occurs over a short period (Ricker type 1) 
#' or is extended over the year and competes with natural mortality and tag
#' sheddding (Ricker type 2) 
#' @param tag_mort initial tag-induced mortality
#' @param reporting vector of tag reporting rate
#' @param nat_mort vector of natural mortality (instantaneous)
#' @param chronic_shed vector of chronic (ongoing) tag shedding
#' @param chronic_mort vector of chronic (ongoing) tag-induced mortality
#' @export
#' @examples 
#' ## Chapman estimates of population numbers
#' single_release(tags=100, catch=200, recaps=5, method="Chapman",
#'                type=1, unit="numbers")
#' single_release(tags=100, catch=200, recaps=5, method="Chapman", 
#'                type=1, unit="numbers", tag_mort=0.1)
#' single_release(tags=100, catch=200, recaps=5, method="Chapman", type=1,
#'                unit="numbers", tag_mort=0.1, reporting=0.8) 
#' 
#' ## Chapman estimate of the population size of carp in a lake
#' ## Load the carphauls data
#' attach(carphauls)
#' fit <- single_release(tags = 803, catch = carphauls$catch,
#'                      recaps = carphauls$tagged,
#'                      prior_recaps = c(108, 181, 82),
#'                      method = "Chapman", unit = "numbers",
#'                      type = 2, tag_mort = 0.3, nat_mort = 0.04,
#'                      chronic_mort = 0.17)
#' summary(fit)
#' 
#' ## Not run: 
#' #######################################
#' # estimate uncertainty with a bootstrap
#' #######################################
#' 
#' boot_fit <- bootstrap(fit, 1000)
#' summary(boot_fit)
#' 
#' ## End(Not run)
single_release <- function(tags, catch, recaps, mean_wt=0, prior_recaps=0, 
                           method, unit, type, tag_mort=0, reporting=1, 
                           nat_mort=0, chronic_shed=0, chronic_mort=0)  {
  ## check the releases are > recaptures
  if((sum(recaps)+sum(prior_recaps)) > tags) 
    stop("more tagged individuals have been recaptured than were released")
  ## check the method
  if(!method %in% c("Petersen", "Chapman")) 
    stop("incorrect method, currently 'Petersen' and 'Chapman' are implemented")
  ## check units
  if(!unit %in% c("numbers", "kg", "tonnes")) 
    stop("incorrect units must be numbers, kg or tonnes")
  ## so we expect to have more than one
  if(length(catch) != length(recaps)) 
    stop("catch and recaptures must be of the same length")
  ## more checks as they come to mind
  ## if catch and recaps are length 1 we can't bootstrap, 
  ## however this isn't necessarily a problem
  ## define the number of years and check vector lengths
  n_years <- length(prior_recaps)
  ## if some parameters are scalars create vectors with rep
  if(n_years > 1){
    ## generalise with a new function
    if(length(reporting)==1) reporting <- rep(reporting, n_years)
    if(length(nat_mort)==1) nat_mort <- rep(nat_mort, n_years)
    if(length(chronic_shed)==1) chronic_shed <- rep(chronic_shed, n_years)
    if(length(chronic_mort)==1) chronic_mort <- rep(chronic_mort, n_years)
  }
  ## check inputs are correct length, again replace with a function
  if(any(length(reporting)!=n_years, length(nat_mort)!=n_years, 
         length(chronic_shed)!=n_years, length(chronic_mort)!=n_years)) 
    stop("Inputs are not of the same length")  
  ## calculate the adjusted numbers
  ## if there are zero recaptures add an NA
  if(sum(recaps)==0){
    ## if there are no recaptures we don't estimate population size
    N_hat=NA
    var_N=NA
    est<- c(N_hat,var_N)
    names(est)<-c("N_hat","var_N")
  }else{
    ## loop over the years to calculate 
    for(i in 1:n_years){
      if(type==1){
        ## calculate the adjusted releases based on tag-induced mortality
        if(i==1){
          ## apply tag-induced mortality in the first year only
          adj_tags <- tags*exp(-tag_mort)
          }
          ## remove recaptures scaled for reporting rate (could improve)
          adj_tags <- adj_tags - (prior_recaps[i] / reporting[i])
          ## apply natural mortality etc
          adj_tags <- adj_tags*exp(-nat_mort[i]-chronic_shed[i]-chronic_mort[i])
      }else if(type==2){
        ## calculate the adjusted releases based on tag-induced mortality
        if(i==1){
          ## apply tag-induced mortality in the first year only
          adj_tags <- tags*exp(-tag_mort)
        }
        ## remove half recaptures scaled for reporting rate (could improve)
        adj_tags <- adj_tags - 0.5*(prior_recaps[i] / reporting[i])
        adj_tags <- adj_tags*exp(0.5*(-nat_mort[i]-chronic_shed[i]-chronic_mort[i]))
        adj_tags <- adj_tags - 0.5*(prior_recaps[i] / reporting[i])
        adj_tags <- adj_tags*exp(0.5*(-nat_mort[i]-chronic_shed[i]-chronic_mort[i]))
      }else stop("type of fishery is not defined")
    ## adjust recaptures in the survey season by dividing by reporting rate 
    adj_recaps <- sum(recaps) / reporting[n_years]
    ## now check that the adjusted releases are > 0
    if(adj_tags <= 0){
      ## alternately N_hat could be zero
      N_hat <- NA 
      var_N <- NA
      est<-c(N_hat,var_N)
      names(est)<-c("N_hat","var_N")
      warning("predicted releases are less than zero based on other parameters")
    }else if(method=="Petersen"){
      ## Petersen numbers and weight are the same
      est <- petersen(adj_tags, sum(catch), adj_recaps)
    }else if(method=="Chapman" & unit %in% c("numbers")){
      ## Chapman numbers
      est <- chapman_n(adj_tags, sum(catch), adj_recaps)
    }else if(method=="Chapman" & unit %in% c("kg", "tonnes")){
      ## Chapman weight
      est <- chapman_wt(adj_tags, sum(catch), adj_recaps, mean_wt)
    }else stop("method and unit combination not available")
    }
  }
  ## store the data
  ###*******************************************###
  hauls <- data.frame(cbind(catch, recaps))
  ## wrap the results in a list
  obj <- list("Estimate" = est,
              "Method" = method,
              "Unit" = unit,
              "Type" = type,
              "Releases" = tags,
              "Hauls" = hauls,
              "MeanWeight" = mean_wt,
              "PriorRecaps" = prior_recaps,
              "TagMort" = tag_mort, 
              "Reporting" = reporting,
              "NatMort" = nat_mort,
              "ChronicShed" = chronic_shed,
              "ChronicMort" = chronic_mort,
              "TagsAvailable" = adj_tags)
  ## add class 'srelease'
  class(obj) <- 'srelease'
  obj
}

## ************************************************ ##
## S3 Generics

#' @export
#' @rdname bootstrap
bootstrap.srelease <- function(x, nboot, boot_zeroes=TRUE, ...){
  ## check the there are sufficient rows in the data
  if(nrow(x$Hauls) <= 1) 
    stop("must be more than one haul to undertake bootstrap")
  if(nrow(x$Hauls) < 10) 
    warning("there are less than 10 hauls used in the bootstrap")
  ## extract the various components of x
  tags <- x$Releases
  hauls <- x$Hauls
  prior_recaps <- x$PriorRecaps 
  tag_mort <- x$TagMort
  reporting <- x$Reporting
  nat_mort <- x$NatMort
  chronic_shed <- x$ChronicShed
  chronic_mort <- x$ChronicMort
  ## extract the number of years to loop over
  n_years <- length(prior_recaps)
  ## object to store the bootstrapped estimates
  boot_est <- data.frame(matrix(NA, nrow=nboot, ncol=3))
  names(boot_est) <- c("boot_est", "boot_var", "boot_recaps")
  ## loop over the number of simulations 
  for(j in 1:nboot){
    ## stochastically apply tag loss and mortality by season
    for(k in 1:n_years){
      ## in the first year we apply tag induced mortality
      if(k==1){
        adj_tags <- rbinom(n=1, size=tags, prob=(exp(-tag_mort)))
      }
      if(x$Type == 1){
        ## remove the recaptures scaled for reporting rate and rounding up
        adj_tags <- ceiling(adj_tags - (prior_recaps[k] / reporting[k]))
        ## apply tag shedding and natural mortality 
        adj_tags <- rbinom(n=1, size=adj_tags, 
                           prob=exp(-nat_mort[k]-chronic_shed[k]-chronic_mort[k]))
      }else if(x$Type == 2){
        ## remove half recaptures scaled for reporting rate and rounding up
        adj_tags <- ceiling(adj_tags - 0.5*(prior_recaps[k] / reporting[k]))
        ## apply half tag shedding and natural mortality 
        adj_tags <- rbinom(n=1, size=adj_tags, 
                           prob=exp(0.5*(-nat_mort[k]-chronic_shed[k]-chronic_mort[k])))
        ## remove second half of the recaptures
        adj_tags <- ceiling(adj_tags - 0.5*(prior_recaps[k] / reporting[k]))
        ## apply half tag shedding and natural mortality 
        adj_tags <- rbinom(n=1, size=adj_tags, 
                           prob=exp(0.5*(-nat_mort[k]-chronic_shed[k]-chronic_mort[k])))
      }else stop("type of fishery is not defined")
    }
    ## check adj_tags >=1 otherwise don't estimate population size and return NA
    if(adj_tags <1){
      boot_est[j,] <- c(NA, NA, NA)
    }else{
      ## resample the haul data 
      j_sample <- sample(nrow(hauls), replace=TRUE)
      j_recaps <- sum(hauls[j_sample,]$recaps)/ reporting[n_years]
      j_catch  <- sum(hauls[j_sample,]$catch)
      ## calculate Chapman estimate
      if(x$Method=="Petersen"){
        ## Petersen numbers
        boot_est[j,] <- c(suppressWarnings(petersen(adj_tags, j_catch, j_recaps)), j_recaps)
      }else if(x$Method=="Chapman" & x$Unit %in% c("numbers")){
        ## Chapman numbers
        boot_est[j,] <- c(suppressWarnings(chapman_n(adj_tags, j_catch, j_recaps)), j_recaps)
      }else if(x$Method=="Chapman" & x$Unit %in% c("kg", "tonnes")){
        ## Chapman weight
        boot_est[j,] <- c(suppressWarnings(chapman_wt(adj_tags, j_catch, j_recaps, x$MeanWeight)), j_recaps)
      }
    }
  }
  ## store the inputs and results 
  obj <- list("srelease_obj" = x,
              "Boot_estimates" = boot_est)
  ## add a class
  class(obj) <- "bsamples"
  ## return the results
  return(obj)
}

#' Print method for single release, single recapture models
#'
#' S3 method for single release, single recapture models 
#' tag recapture models of abundance
#' @param x object of class srelease
#' @param ... additional parameters
#' @export
print.srelease <- function(x, ...){
  ## define the model used
  cat(x$Method, "estimate of population", x$Unit, "\n")
  print(x$Estimate)
}

#' Summary method for single release, single recapture models
#'
#' S3 method for single release, single recapture models 
#' tag recapture models of abundance
#' @param object object of class srelease
#' @param ... additional parameters
#' @export
summary.srelease <- function(object, ...){
  x <- object
  ## print the model inputs 
  cat(x$Method, "estimate of", x$Unit, "\n")
  print(x$Estimate)
  cat("With cv", sqrt(x$Estimate["var_N"]) / x$Estimate["N_hat"], "\n")
  ## summarise the inputs
  cat(x$Releases, "tags were released and", sum(x$PriorRecaps),
      "were subsequently recaptured prior to this season \n")
  cat(sum(x$Hauls["catch"]), x$Unit, "were captured in the current survey and",
      sum(x$Hauls["recaps"]), "were tagged \n")
  cat("The following parameters were specified \n")
  cat("Initial tag-induced mortality =", x$TagMort, "\n")
  cat("Tag reporting rates by season =", x$Reporting, "\n")
  cat("Natural mortality by season =", x$NatMort, "\n")
  cat("Chronic tag shedding by season =", x$ChronicShed, "\n")
  cat("Chronic tag-induced mortality by season =", x$ChronicMort, "\n")
  cat("Tags available by season =", x$TagsAvailable,"\n")
}

#' S3 method for bootstrapped confidence intervals
#'
#' Calculate bootstrapped confidence intervals for object of
#' class bsamples
#' @param object object of class bsamples
#' @param quantiles bootstrap quantiles (default 0.025, 0.5, 0.975)
#' @param ... additional parameters
#' @export
summary.bsamples <- function(object, quantiles=c(0.025, 0.5, 0.975), ...){
  ## define the quantiles
  quants <- quantile(object$Boot_estimates$boot_est, probs=quantiles, na.rm=TRUE)
  se <- sd(object$Boot_estimates$boot_est, na.rm=TRUE)
  cv <- se / object$srelease_obj$Estimate["N_hat"]
  names(se) <- "boot_SE"
  names(cv) <- "boot_CV"
  ## construct the output
  out <- c(object$srelease_obj$Est[1], se, cv, quants)
  #names(out) <- c("Estimate", "lower", "upper")
  out
}
